sdks/go/pkg/beam/core/metrics/metrics.go

```diff
@@ -448,3 +453,73 @@ func (m *gauge) get() (int64, time.Time) {
 	defer m.mu.Unlock()
 	return m.v, m.t
 }
+
+// GaugeValue is the value of a Gauge metric.
+type GaugeValue struct {
+	Value     int64
+	Timestamp time.Time
+}
+
+// Results represents all metrics gathered during the job's execution.
+// It allows for querying metrics using a provided filter.
+type Results struct {
+	counters      []CounterResult
+	distributions []DistributionResult
+	gauges        []GaugeResult
+}
+
+// AllMetrics returns all metrics from a Results instance.
+func (mr Results) AllMetrics() QueryResults {
+	return QueryResults{mr.counters, mr.distributions, mr.gauges}
+}
+
+// TODO(BEAM-11217): Implement Query(Filter) and metrics filtering
+
+// QueryResults is the result of a query. Allows accessing all of the
+// metrics that matched the filter.
+type QueryResults struct {
+	counters      []CounterResult
+	distributions []DistributionResult
+	gauges        []GaugeResult
+}
+
+// GetCounters returns an array of counter metrics.
+func (qr QueryResults) GetCounters() []CounterResult {
```

Comment:
Perhaps we could skip these getters and expose QueryResults's fields directly. We don't need to follow Java API that much. @lostluck What do you think?

Reply:
Let's keep these methods, but remove the Get* prefix.

Regarding the results, a safer option is probably to return their copies instead. The structs don't have any underlying pointers, so it's not a problem to use built-in `copy` function.